### DISCUSSION TITLE
How can I save the built-in storage as multi-layered JSON?

### DISCUSSION THREAD

**Discussion by Aues6uen11Z** [zauberzeug/nicegui#2923] [Q&A]
### Question

Hi, In my case, I need to use app.storage.general to bind the value of a element, but there are also other places that need to access these values. I am using it like this:
``` py
ui.input('test').bind_value(app.storage.general, 'test')
```
However, this usage only allows for a single layer of key:value mapping. If I try to use it like this:
```py
ui.input('test').bind_value(app.storage.general['a'], 'test')
```
It will report a KeyError because the key 'a' does not exist. In order to easily access these values in other places, I expect to be able to build a multi-layered JSON data structure. Is there a way to do this now?

**Comment by falkoschindler** [comment]
Hi @Aues6uen11Z,
I think this isn't possible. Even though the attribute "test" can still be undefined when calling `bind_value`, the object you're binding to or from needs to exist.

**Reply by Aues6uen11Z** [reply]
Hi, thank you for your reply. Is it possible for NiceGUI to introduce a special dictionary that can create key values at any depth?

```python
from collections import defaultdict
from typing import Dict


class InfDict(defaultdict):
    """Infinite nested dict, you can create keys at any depth without worrying about KeyError"""
    def __init__(self, common_dict: Dict = None):
        super().__init__(self.__class__)
        if common_dict:
            if isinstance(common_dict, dict):
                self.update({k: self.__class__(v) if isinstance(v, dict) else v for k, v in common_dict.items()})
    
    def __repr__(self):
        return dict(self).__repr__()

test = InfDict()
test['a']['b']['c'] = 1
print(test)
# {'a': {'b': {'c': 1}}}
```
If the dict of the `PersistentDict` class can be replaced with this, it should achieve the effect I want, but I am not sure if this will compromise the reliability of the program and bring other issues.

**Reply by falkoschindler** [reply]
As far as I know, [addict](https://github.com/mewwts/addict) does something similar, as discussed in this feature request: https://github.com/zauberzeug/nicegui/discussions/1680
But I'm a bit hesitant to change NiceGUI's storage dictionaries to addict or similar. While providing a simpler syntax in some cases, it might hide bugs in other cases. It will silently return `None` (I guess) even if the key is wrong.

**Reply by Aues6uen11Z** [reply]
You are right, if using the InfDict I wrote above, printing a non-existent key will result in an empty dictionary “{}”. However, I suspect that most people who use the built-in storage functionality do so to bind values rather than explicitly using the PersistentDict object, right? If it’s only used to bind component values, and the accessed keys are fixed, there shouldn’t be any issues. In any case, there are indeed pros and cons, and the decision is yours. If it won't change, I can achieve my goal in a different way, Thanks.

**Comment by Aues6uen11Z** [comment]
Hi, @falkoschindler 
I have a new idea, perhaps we don’t need to modify the `PersistDict`, but only change `bind_value`. We can manually create those non-existent keys during binding, like this.
```python
# nicegui/elements/mixins/value_element.py


def _convert_target(
        self,
        target_object: Any,
        target_name: Union[str, Tuple[str, ...]]
        ) -> Tuple[Any, str]:
    if isinstance(target_name, tuple):
        if isinstance(target_object, dict):
            for key in target_name[:-1]:
                try:
                    target_object[key]
                except KeyError:
                    target_object[key] = {}
                target_object = target_object[key]
            target_name = target_name[-1]
        else:
            raise TypeError

    return target_object, target_name

def bind_value(self,
                target_object: Any,
                target_name: Union[str, Tuple[str, ...]] = 'value', *,
                forward: Callable[..., Any] = lambda x: x,
                backward: Callable[..., Any] = lambda x: x,
                ) -> Self:

    # Also in bind_value_to, bind_value_from
    target_object, target_name = self._convert_target(target_object, target_name)
    bind(self, 'value', target_object, target_name, forward=forward, backward=backward)
    return self
```
Then we can use it like this, and it won’t affect the original usage.
```python
from nicegui import ui
from nicegui.storage import PersistentDict


test = PersistentDict(Path('test.json'))

ui.input().bind_value(test, ('1', '2', '3'))
ui.input().bind_value(test, ('4', '5'))
ui.input().bind_value(test, ('6'))
ui.run()
"""
test.json:

{
    "1": {
        "2": {
            "3": "test"
        }
    },
    "4": {
        "5": "test"
    },
    "6": "test"
}
"""
```
This way, we can create structured JSON data for persistent storage and also trigger a KeyError when there is an incorrect access. What do you think?

**Reply by falkoschindler** [reply]
Thanks for the suggestion, @Aues6uen11Z. We are still a bit sceptical about extending the existing binding API in this way. But we see the potential for simplifying user code. Would you like to create a [feature request](https://github.com/zauberzeug/nicegui/discussions/new?category=ideas-feature-requests) so that we can discuss its pros and cons with the community and collect upvotes?

**Reply by Aues6uen11Z** [reply]
Sure, I made a new discussion in https://github.com/zauberzeug/nicegui/discussions/2978
